### PR TITLE
deployer: create separate new typer apps under the deployer and update the general structure of the directory

### LINKED ISSUE #971: Refactor for clarity that deployer/tests aren't just developer tests
## Description of problem and opportunity to address it

The `deployer` CLI rely on `pytest` as a production dependency - not just development dependency. This is because it use `pytest` and the tests in `deployer/tests` as a way to verify that hub's are functional.

From a glance, I would assume that those tests would be tests associated with testing the `deployer` CLI itself, because that is a pattern I've seen across the python ecosystem repeated. This is the first time I see `pytest` being used a real requirement for a project. Due to that, I suggest we are more explicit about this somehow.

### Should pytest be used at all?
@sgibson91 mentioned to me that @yuvipanda wants to avoid using `pytest` as a production dependency. That sounds reasonable to me and would solve this issue as well.

### Pytest and logging
@sgibson91 also noted that there has been issues with a hub health-check test that fails could lead to `pytest`'s verbosity exposes sensitive credentials. By doing a test manually, we would also avoid such complexity. This is related to:
- https://github.com/2i2c-org/infrastructure/issues/585
- https://github.com/2i2c-org/infrastructure/issues/694.

## Updates and ongoing work

- 2023-10-11, this issue is still relevant but low prio.

### PR DESCRIPTION
**Pull request by GeorgianaElena**
Fixes https://github.com/2i2c-org/infrastructure/issues/2024

## What the PR does
This PR refactors the deployer's directory structure and creates new  typer applications under the main `deployer` app.
The commands that we have, then register under these new applications based on their location in the package.

*NOTE*: the only new code that this PR adds is related to the creation of these typer apps. Other then this, this mostly is about moving files around, renaming them and extracting functions into their own files. More in the  commits' description.

## Benefits
We are now left with a more clear to follow structure of the main `deployer` directory instead of having an endless number of python files in a big pool. 
More, this directory structure maps to the deployer commands, so that it's easier to know where the code that describes a command is.

## A note to review more easily
I was very grateful each time @consideRatio split his work into individual logical commits and described what each commit did. That greatly improved my understanding of what the PR did and presented the changes gradually, hence the review was easier <3.

Although I didn't start with this thought in mind because this PR was supposed to be something else and smaller, I did my best to squash the relevant ones together and provide a description for each of them.

So, to review more easily, I suggest going to https://github.com/2i2c-org/infrastructure/pull/3094/commits and look at each commit and its description at a time. Thank you!!!

---

EDIT: fixes #971

### CODE REVIEW

File: deployer/README.md
Comment (yuvipanda) [inline deployer/README.md:63]: Given we're already in a subdirectory called `commands`, I think it's ok to not have this additional suffix. Less clutter that way.

File: deployer/README.md
Comment (yuvipanda) [inline deployer/README.md:102]: While we're moving this around, let's call these 'health-check-tests' or something like that for clarity, that it's not testing the deployer itself?

File: deployer/README.md
Comment (yuvipanda) [inline deployer/README.md:695]: Let's invert this, so `debug` is the top level, and `exec` is a sub-level? None of the things under `debug` are execing anything, while usually we `exec` things to debug something...

Comment (yuvipanda) [reply deployer/README.md:695]: So the layout would be:

```
- debug
  |- exec
  |- user-logs
  |- component-logs
  |- start-docker-proxy
```

Comment (yuvipanda) [reply deployer/README.md:695]: Alternatively, `debug` and `exec` can *both* be top level.

File: deployer/infra_components/cluster.py
Comment (yuvipanda) [inline deployer/infra_components/cluster.py:18]: I feel this is generic enough it should continue to be in something under `utils` rather than here.

**Review by yuvipanda — CHANGES_REQUESTED** [review]
Thanks @GeorgianaElena! Generally looks good to me, made some minor suggestions.

**Review by consideRatio — COMMENTED** [review]
I love that tests is being renamed to involve health_checks, i consider that to fix #971!

**Comment by GeorgianaElena** [comment]
> I love that tests is being renamed to involve health_checks, i consider that to fix https://github.com/2i2c-org/infrastructure/issues/971!

This was part of @yuvipanda's feedback <3 🚀

> Thanks @GeorgianaElena! Generally looks good to me, made some minor suggestions.

I've udpated based on your feedback, Yuvi and rebased & fix the conflicts.

File: deployer/infra_components/cluster.py
Comment (consideRatio) [inline deployer/infra_components/cluster.py:15]: I expected `from .hub` to become `from deployer.infra_components.hub` given the other changes so that we consistently use either relative or absolute references.

I'm generally a bit uncertain about the choice between relative `.` references and absolute `package_name` references, not knowing the consequences of going for one, the other, or a mix.

**Review by consideRatio — COMMENTED** [review]
The UI is far better with the sub-commands for sure, nice!

I expect this to be safe to merge for our production hubs operation since we have staging hubs also. I've not reviewed all commits or all changed lines, but skimmed through a bit.

To resolve:
- another merge conflict
- comment about `from .hub` import

File: deployer/infra_components/cluster.py
Comment (GeorgianaElena) [reply deployer/infra_components/cluster.py]: @consideRatio, the main motivation for switching the majority of imports from relative to absolute, is because with this PR at least 2 or 3 more levels are added and it relative imports were not serving their purpose anymore when you have something like `....file_acquisition import find_absolute_path_to_cluster_file`

(Also, apparently there's a PEP that suggests using absolute ones https://peps.python.org/pep-0008/#imports)

But I believe relative imports are still useful for the `deployer` package when referring to the current directory.

For example, in `deployer/commands/generate/dedicated_cluster/aws.py` we have a very simple line that says:
```
from .dedicated_cluster_app import dedicated_cluster_app
```
if we turn it into an absolute import then it becomes:
```
from deployer.commands.generate.dedicated_cluster.dedicated_cluster_app import dedicated_cluster_app`
```

Although not recommended (relative imports are not recommended anyhow), I would leave these imports relative to the current directory be, for simplicity. @consideRatio, do you feel strongly about using absolute imports everywhere? If yes, I will update them.

I added a commit that uses absolute imports in a few more places, but haven't touched the ones in the `commands` directory for the reason above.

Comment (consideRatio) [reply deployer/infra_components/cluster.py]: > do you feel strongly about using absolute imports everywhere? If yes, I will update them.

No, I'm mostly cautious because I recall issues that I didn't fully understand related to using one or the other kind.

> leave these imports relative to the current directory be

Sure!

**Review by consideRatio — APPROVED** [review]
Thank you for working this @GeorgianaElena!! Happy to see sub-commands adopted to compartmentalize the UI and code behind it!

I think we should review this further by testing it rather than by further code review. I don't expect this to break things in a way that we don't catch.

@GeorgianaElena maybe you could do a deploy-support and deploy a staging hub before merging though as a precaution, with a health check on the staging before we trigger the full deploy?

**Comment by GeorgianaElena** [comment]
> @GeorgianaElena maybe you could do a deploy-support and deploy a staging hub before merging though as a precaution, with a health check on the staging before we trigger the full deploy?

Just ran them locally with no issue + a validate command. 

> I think we should review this further by testing it rather than by further code review. I don't expect this to break things in a way that we don't catch.

Thank you @consideRatio! Merging this now 🚀

**Pull request merged**
